src/VisualStudio/Core/Def/Implementation/Utilities/ServiceInitializer.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Roslyn.Utilities;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Utilities
+{
+    internal class ServiceInitializer<TInterface, TService> where TInterface : class
+    {
+        private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
```

Comment:
This will deadlock when a background thread obtains the semaphore and tries to switch to the UI thread, while the UI thread is calling `GetService` on this class and is blocked waiting for the semaphore.
You need to follow the patterns of the 3rd threading rule for this. As this is the most tricky rule to follow when you have hidden queues like this, we've just recently encoded it in a JTF-aware semaphore class! Using `Microsoft.VisualStudio.ReentrantSemaphore` here should work for you.

Reply:
so. you mean if I do JaoinabletaskFactory.Run(() => SwitchToUIThreadAsync()) on UI thread, it will deadlock? so unlike Task.Run, JoainableTaskFactory.Run will try to run Func at the same thread as it is called?

Reply:
@AArnott if I make await switchtouithread outside of lock, will that make it not to deadlock?

Reply:
hmmm... how can I make async method to always run on BG and switch backto UI thread and go back to BG?

Reply:
> so. you mean if I do JaoinabletaskFactory.Run(() => SwitchToUIThreadAsync()) on UI thread, it will deadlock?

Not generally, no. That pattern is fine. But if you do that while holding a private lock/semaphore of your own, you'll still be holding that while trying to switch to the UI thread. And if the UI thread is synchronously blocked waiting for that same lock, then you have yourself the classic salt+pepper deadlock.
The `ReentrantSemaphore` class is specifically written to mitigate that deadlock by having the party controlling the UI thread let the lock holder in so it can complete its work in order to release the lock so the UI thread can resume its own work. 

> so unlike Task.Run, JoainableTaskFactory.Run will try to run Func at the same thread as it is called?

Correct: JTF.Run synchronously blocks while JTF.RunAsync does not. JTF.RunAsync *invokes* the delegate inline with its caller, but when it yields, so does JTF.RunAsync. Thus, JTF *never* does a thread transition for you implicitly -- it's always an explicit request of yours. Task.Run of course never executes the delegate inline, and always runs the delegate on a threadpool thread.

> if I make await switchtouithread outside of lock, will that make it not to deadlock?

Correct: if you aren't holding a lock while transitioning to the UI thread, it shouldn't deadlock with the UI thread that is trying to get the lock -- you'll just block until the UI thread is available. 

> how can I make async method to always run on BG and switch backto UI thread and go back to BG?

An async method can switch to the background thread with `await TaskScheduler.Default;` then switch to the UI thread with `await JTF.SwitchToMainThreadAsync()`. You can use these two as frequently as you need in an async method.

Reply:
@AArnott either my english is wrong or we have some misunderstanding. if JTF.Run(... delegate ) runs the delegate in threadpool, then current code should be fine. UI thread never holds the private lock. GetService doesn't acquire lock only GetServiceAsync does. and it does before switching to UI thread. so when JTF switch to UI thread under Run, there should be no deadlock situation.

Reply:
>  if JTF.Run(... delegate ) runs the delegate in threadpool, then current code should be fine.

JTF.Run invokes the delegate on its caller's thread. It does *not* switch to the thread pool. If the delegate itself switches threads (whether to the threadpool or the main thread) then at that point *two* threads are occupied: the original thread that is JTF.Run blocking on completion of the delegate, and the thread that the delegate has switched to internally.

Reply:
@AArnott I see. I assumed it will run the delegate on BG. anyway, good to know.